### PR TITLE
Implement Response abstraction for Azure

### LINKED ISSUE #45705: Update ClientResult to Azure.Core.Response for Azure plugin
### Library name

N/A

### Please describe the feature.

Update operation return type from `ClientResult` to `Azure.Core.Response`

### PR DESCRIPTION
**Pull request by live1206**
Resolves #45705

### CODE REVIEW

**Comment by live1206** [comment]
For now, we can't consume NuGet package from MGC due to https://github.com/microsoft/typespec/issues/4507.

File: eng/packages/http-client-csharp/generator/Azure.Generator.sln
Comment (live1206) [inline eng/packages/http-client-csharp/generator/Azure.Generator.sln:8]: This should be reverted before merging, for temp local test purpose.

File: eng/packages/http-client-csharp/generator/Azure.Generator/src/Azure.Generator.csproj
Comment (live1206) [inline eng/packages/http-client-csharp/generator/Azure.Generator/src/Azure.Generator.csproj:16]: This should be reverted before merging, for temp local test purpose.

**Comment by live1206** [comment]
Will create a new PR to upgrade MGC version to reduce the changes in this PR.

**Comment by jsquire** [comment]
It would be super helpful for these to see an example of what the generated code looks like.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.cs
Comment (live1206) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.cs:59]: @jsquire 
You can find the generated SDK for Azure under this test project.
For instance, the Pipeline type has been changed from `ClientPipeline` to `HttpPipeline`.
And the response type of operation has been changed from `ClientResult` to `Response`, and such on.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs
Comment (pshao25) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs:308]: Just curious, shouldn't it be Classifier200And204 and only accept 200 and 204? I assume all the code should be the same as what we previous generated in autorest.csharp, is that still makes sense?

Comment (live1206) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs:308]: actually, this should be omitted here

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs
Comment (live1206) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs]: fixed parameter name, it is in use though

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs
Comment (live1206) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs:308]: This needs further tweak in MGC, I will leave it to next iteration as the change in MGC is pretty big already.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs
Comment (pshao25) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs]: Huh? Did I miss anything? I didn't see it attached to message.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/BasicTypeSpec.csproj
Comment (ArcturusZhang) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/BasicTypeSpec.csproj:15]: we need a format here

Comment (live1206) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/BasicTypeSpec.csproj:15]: hmm... this is auto-generated, and the format is correct in vscode and VS.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs
Comment (live1206) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs]: I will resolve this in https://github.com/microsoft/typespec/pull/4816

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/BasicTypeSpec.csproj
Comment (pshao25) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/BasicTypeSpec.csproj:15]: It's two tabs before this line. Others are 4 spaces.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/BasicTypeSpec.csproj
Comment (pshao25) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/BasicTypeSpec.csproj:13]: I saw you deleted a lot of using for ClientModel. Then do we still need this reference?

Comment (live1206) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/BasicTypeSpec.csproj:13]: because not everything has been abstracted, such as `ApiKeyCredential` will be handled later, then we can remove this.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/BasicTypeSpec.csproj
Comment (live1206) [reply eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/BasicTypeSpec.csproj:15]: And this is actually manually added for now, will need to generate it.

**Review by pshao25 — APPROVED** [review]

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.cs
Comment (annelo-msft) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.cs:55]: Should this have an authentication policy?

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs
Comment (annelo-msft) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs:32]: RequestMethod.Get will prevent making an allocation here.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs
Comment (annelo-msft) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs:42]: Why Add and not Set?

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs
Comment (annelo-msft) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClient.RestClient.cs:33]: Does Azure.Core have ClientUriBuilder?  I believe Azure clients can set the URI in the request directly.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClientOptions.cs
Comment (annelo-msft) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/BasicTypeSpecClientOptions.cs:14]: These will be more different going forward ... just a heads up that more will need to differ here.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/Internal/ClientPipelineExtensions.cs
Comment (annelo-msft) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/Internal/ClientPipelineExtensions.cs:19]: This should pass the CancellationToken through to pipeline.Send from RequestContext.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/Internal/ClientPipelineExtensions.cs
Comment (annelo-msft) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/Internal/ClientPipelineExtensions.cs:26]: We should sync on ExtractResponseContent ... this may not be the right pattern for Azure clients.

File: eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/Internal/ErrorResult.cs
Comment (annelo-msft) [inline eng/packages/http-client-csharp/generator/TestProjects/Local/Basic-TypeSpec/src/Generated/Internal/ErrorResult.cs:13]: We should remove this from unbranded clients.  SCM-based clients should never throw from a property.

**Comment by live1206** [comment]
@annelo-msft Thanks a lot for reviewing the generated SDK for Azure!
The idea is not to have everything in place in this PR.
I will create issue to track your comments and follow-up PRs to mitigate the remaining issues.

**Comment by annelo-msft** [comment]
Sounds great.  Many thanks, @live1206!

**Pull request merged**
